### ISSUE TITLE
Caching pipes with binary data

### ISSUE THREAD

**Issue by enricodeleo** [evanshortiss/express-expeditious#14]
Hi!

First of all thank you for this gear, very clever. For me everything seems working but I'm having bad times trying to cache a `.pipe(res)` coming from [node canvas](https://github.com/Automattic/node-canvas). I think the culprit could be being a binary stream, couldn't it? 
At first I tried adding `detect_buffers: true` among redis engine options but nothing changed. I then switched to memory engine (I was able to find the cached key/value of the request on redis by the way) but I have the same result so I thought it could be the way the middleware itself treats data. Any ideas?

**Comment by evanshortiss** [comment]
Yeah, sounds like it could be something to do with binary vs. text streams. Could you create a really simple example to reproduce the issue and link it here?

**Comment by enricodeleo** [comment]
Thank you for your reply @evanshortiss. I made up this simple proof of concept https://github.com/enricodeleo/express-pipe-cache-test/

As you will see there are 4 endpoints
`/` it's a simple res.send with an index of the other 3 endpoints
`/json-works` pipes a response from a json api that works as expected 
`/html-fails` that's kinda surprise because it's something within your example directory (well, you fetches facebook.com and I'm fetching google.com :D) and here I have the same behaviour I was expecting with binary (I cannot see a response when hitting the cache)
`/gif-fails` here the issue I was referring to. A stream of an image created with gifencoder and node-canvas.

Now since I'm facing the same issue piping a webpage (google.com) I'm thinking maybe it is caused somehow by the length of the payload the cache have to cache (json data works because is lighter than an html page).

I tried under node 8 and 9 on MacOS both with memory engine and redis engine.

**Comment by etx** [comment]
I'm having the same problem when caching images resized with the sharp module. 

```
fs.createReadStream('./site/uploads/'+r[0].url)
        .on('error', function(error) {
          next(error)
        })
        .pipe(sharp().resize(30, 30).max().toFormat('webp'))
        .pipe(res)
```

**Comment by enricodeleo** [comment]
I ended up caching directly on redis thanks to the libraries cited on my previous comment.

**Comment by evanshortiss** [comment]
@enricodeleo this is interesting. I just checked your application and the`html-fails` endpoint is fine for me. The GIF endpoint is definitely having issues though.

Here's what happens locally for me when I hit your GIF endpoint using cURL. This happens using Redis, or using the default in memory storage of express-expeditious, so I think it might be an issue with the way this module deals with chunked responses or buffers as you say.

<img width="1530" alt="screen shot 2018-05-08 at 11 03 04" src="https://user-images.githubusercontent.com/1303687/39774368-809f5044-52af-11e8-8a1d-820c65d497bc.png">

Did you find anything else when debugging? I'll need to look into the code but reckon the treatment of all data as plain strings is the culprit.

**Comment by evanshortiss** [comment]
@enricodeleo ok, so by updating the buffer/write code in this module I think I've sorted out part of the issue, I now get consistently sized outputs via cURL:

<img width="1550" alt="screen shot 2018-05-08 at 11 56 52" src="https://user-images.githubusercontent.com/1303687/39776910-0018dfd2-52b7-11e8-82aa-e2c494caf2d2.png">

Looks like this next part to solve might be related to #17. Binary data is not being parsed correctly.

**Comment by evanshortiss** [comment]
@enricodeleo @etx can you try testing against the branch this PR is based on https://github.com/evanshortiss/express-expeditious/pull/19/files 

This _should_ fix things up.

**Comment by etx** [comment]
Just gave it a quick and dirty test. So far it's working as expected. Thanks so much for the quick fix! If I find anything else unexpected I will report back.

**Comment by evanshortiss** [comment]
Awesome @etx, I'll release it via npm soon. I'm considering also changing the structure of the cached object a little to make it more structured since right now it's just two text blobs, but the primary problem seems is resolved so we're good for now.

Thanks for confirming 👍

**Comment by enricodeleo** [comment]
@evanshortiss thank you for the great work! I eventually had the time to switch from my previous custom implementation with `redis-rstream` and `redis-wstream` back again to express-expeditious and I can confirm it works. 
Please, ping us when you publish on npm so we can update our package.json :)

**Comment by evanshortiss** [comment]
@enricodeleo thanks for testing. I will try get this merged in the next 48 hours 👍

**Comment by evanshortiss** [comment]
@ryanto @etx @enricodeleo published to npm as 4.0.0
